stingray/fourier.py

```diff
@@ -0,0 +1,1286 @@
+
+    Parameters
+    ----------
+    C : complex `np.array`
+        cross spectrum
+    Ps : float `np.array`
+        sub-band periodogram
+    Pr : float `np.array`
+        reference-band periodogram
+    Psnoise : float
+        Poisson noise level of the sub-band periodogram
+    Prnoise : float
+        Poisson noise level of the reference-band periodogram
+    Nave : int
+        number of intervals that have been averaged to obtain the input spectra
+
+    Other Parameters
+    ----------------
+    common_ref : bool, default False
+        Are data in the sub-band also included in the reference band?
+
+    Returns
+    -------
+    dRe : float `np.array`
+        Error on the real part of the cross spectrum
+    dIm : float `np.array`
+        Error on the imaginary part of the cross spectrum
+    dphi : float `np.array`
+        Error on the angle (or phase lag)
+    dG : float `np.array`
+        Error on the modulus of the cross spectrum
+
+    """
+    twoN = 2 * Nave
+    if common_ref:
+        Gsq = (C * C.conj()).real
+        bsq = bias_term(C, Ps, Pr, Psnoise, Prnoise, Nave)
+        frac = (Gsq - bsq) / (Pr - Prnoise)
+        PoN = Pr / twoN
+
+        # Eq. 18
+        dRe = dIm = dG = np.sqrt(PoN * (Ps - frac))
+        # Eq. 19
+        dphi = np.sqrt(PoN * (Ps / (Gsq - bsq) - 1 / (Pr - Prnoise)))
+
+    else:
+        PrPs = Pr * Ps
+        dRe = np.sqrt((PrPs + C.real ** 2 - C.imag ** 2) / twoN)
+        dIm = np.sqrt((PrPs - C.real ** 2 + C.imag ** 2) / twoN)
+        gsq = raw_coherence(C, Ps, Pr, Psnoise, Prnoise, Nave)
+        dphi = np.sqrt((1 - gsq) / (2 * gsq ** 2 * Nave))
+        dG = np.sqrt(PrPs / Nave)
+
+    return dRe, dIm, dphi, dG
+
+
+def cross_to_covariance(C, Pr, Prnoise, delta_nu):
+    """Convert a cross spectrum into a covariance.
+     Parameters
+    ----------
+    C : complex `np.array`
+        cross spectrum
+    Pr : float `np.array`
+        reference-band periodogram
+    Prnoise : float
+        Poisson noise level of the reference-band periodogram
+    delta_nu : float or `np.array`
+        spectral resolution. Can be a float, or an array if the spectral
+        resolution is not constant throughout the periodograms
+
+    """
+    return C * np.sqrt(delta_nu / (Pr - Prnoise))
+
+
+def _which_segment_idx_fun(counts=None, dt=None):
+    # Make function interface equal (counts gets ignored)
+    if counts is None:
+        return generate_indices_of_segment_boundaries_unbinned
+
+    def fun(*args, **kwargs):
+        return generate_indices_of_segment_boundaries_binned(*args, dt=dt, **kwargs)
+    return fun
+
+
+def get_total_ctrate(times, gti, segment_size, counts=None):
+    """Calculate the average count rate during the observation.
+
+    This function finds the same segments that the PDS will use and
+    returns the mean count rate.
+    If ``counts`` is ``None``, the input times are interpreted as events.
+    Otherwise, the number of events is taken from ``counts``
```

Comment:
Oh, so `times` can be *either* time stamps of a binned light curve, *or* arrival times for an unbinned event list? I think it's worth being really explicit about this. 

I'm a bit worried that this could lead to pretty bad errors if the user doesn't think that `counts` is required. Maybe we want two separate functions, one for binned light curves and one for event lists?

Reply:
I tried to make the signatures of the avg_XX_from_iterable functions use the same approach: if you just give times, they are interpreted as events. If you pass a counts array, they are interpreted as light curves instead. This can be done because this and other ancillary functions make this check and give the relevant answers. Splitting this function would just move the control somewhere else. We can discuss that, of course.